design/range-deletion-in-splay-tree.md

```diff
@@ -0,0 +1,80 @@
+---
+title: delete-range-in-splay-tree
+target-version: X.X.X
+---
+---
+
+# Range Deletion in Splay Tree
+
+## Summary
+
+In Splay Tree, an access for a node make the node the root of tree. So when deleting many nodes like `deleteNodes` in `pkg/json/rga_tree_split.go`, deletion for each node in Splay Tree calls many unnecessary `rotation`s.
+
+Using the propertiy of Splay Tree that change the root freely, `splay.DeleteRange` separates the given range to be deleted as a subtree near the root, without rotations of each node.
+
+### Goals
+
+The function `DeleteRange` should separate all nodes exactly in the given range as a subtree. After the function ends, the entire tree from and weight of every node must be correct just as when the nodes were deleted one by one.
+
+## Proposal Details
+
+![delete-range-in-splay-tree-1](./media/range-deletion-in-splay-tree-1-index-of-subtrees.png)
+
+From the property of indexed BST, all nodes with a smaller index than the root are in the left subtree of root and the nodes with a bigger index are in the right subtree.
+
+And also, Splay Tree can change the root freely to use `Splay`.
+
+Then using the properties, when we want to delete the range from index `L` to `R` we can make the shape of tree like the figure avobe to `Splay(L-1)` then `Splay(R+1)`.
+
+
+![delete-range-in-splay-tree-2](./media/range-deletion-in-splay-tree-2-separation.png)
+
+```go
+// see with the figures above
+// leftBoundary is indexed L-1 and rightBoundary is indexed R+1.
+func (t *Tree[V]) DeleteRange(leftBoundary, rightBoundary *Node[V]) {
+    // absence of rightBondary(index R+1) means deletion to the most right of tree.
+	if rightBoundary == nil {
+		t.Splay(leftBoundary)
+		t.cutOffRight(leftBoundary)
+		return
+	}
+
+	t.Splay(leftBoundary)
+	t.Splay(rightBoundary)
+
+    // refer case 2 of second figure
+	if rightBoundary.left != leftBoundary {
+		t.rotateRight(leftBoundary)
+	}
+	t.cutOffRight(rightBoundary)
+}
+
+```
+Sometimes the tree shapes like case 2 after `Splay`s because of the zig-zig case of `Splay`. But it simply changes to the same shapes as case 1 in one rotationfor `L-1`.
```

Comment:
```suggestion
Sometimes the tree shapes like case 2 after `Splay`s because of the zig-zig case of `Splay`. But it simply changes to the same shapes as case 1 in one rotation for `L-1`.
```